extracted/konan.serializer/src/org/jetbrains/kotlin/serialization/konan/impl/KonanResolvedModuleDescriptorsFactoryImpl.kt

```diff
@@ -0,0 +1,181 @@
+import org.jetbrains.kotlin.incremental.components.LookupLocation
+import org.jetbrains.kotlin.incremental.components.NoLookupLocation
+import org.jetbrains.kotlin.konan.library.KonanLibrary
+import org.jetbrains.kotlin.konan.library.resolver.KonanLibraryResolveResult
+import org.jetbrains.kotlin.konan.library.resolver.PackageAccessedHandler
+import org.jetbrains.kotlin.konan.util.profile
+import org.jetbrains.kotlin.name.FqName
+import org.jetbrains.kotlin.name.Name
+import org.jetbrains.kotlin.resolve.descriptorUtil.builtIns
+import org.jetbrains.kotlin.resolve.scopes.MemberScope
+import org.jetbrains.kotlin.resolve.scopes.MemberScopeImpl
+import org.jetbrains.kotlin.serialization.konan.KonanDeserializedModuleDescriptorFactory
+import org.jetbrains.kotlin.serialization.konan.KonanResolvedModuleDescriptors
+import org.jetbrains.kotlin.serialization.konan.KonanResolvedModuleDescriptorsFactory
+import org.jetbrains.kotlin.storage.LockBasedStorageManager
+import org.jetbrains.kotlin.storage.StorageManager
+import org.jetbrains.kotlin.storage.getValue
+import org.jetbrains.kotlin.utils.Printer
+
+class KonanResolvedModuleDescriptorsFactoryImpl(
+        override val moduleDescriptorFactory: KonanDeserializedModuleDescriptorFactory
+): KonanResolvedModuleDescriptorsFactory {
+
+    override fun createResolved(
+            resolvedLibraries: KonanLibraryResolveResult,
+            storageManager: StorageManager,
+            builtIns: KotlinBuiltIns?,
+            specifics: LanguageVersionSettings,
+            customAction: ((KonanLibrary, ModuleDescriptorImpl) -> Unit)?
+    ): KonanResolvedModuleDescriptors {
+
+        val moduleDescriptors = mutableListOf<ModuleDescriptorImpl>()
+        @Suppress("NAME_SHADOWING")
+        var builtIns = builtIns
+
+        // build module descriptors
+        resolvedLibraries.forEach { library, packageAccessedHandler ->
+            profile("Loading ${library.libraryName}") {
+
+                // MutableModuleContext needs ModuleDescriptorImpl, rather than ModuleDescriptor.
+                val moduleDescriptor = createDescriptorOptionalBuiltsIns(
+                        library, specifics, storageManager, builtIns, packageAccessedHandler)
+                builtIns = moduleDescriptor.builtIns
+                moduleDescriptors.add(moduleDescriptor)
+
+                customAction?.invoke(library, moduleDescriptor)
+            }
+        }
+
+        val forwardDeclarationsModule = createForwardDeclarationsModule(builtIns, storageManager)
+
+        // set inter-dependencies between module descriptors, add forwarding declarations module
+        for (module in moduleDescriptors) {
+            // Yes, just to all of them.
+            module.setDependencies(moduleDescriptors + forwardDeclarationsModule)
+        }
+
+        return KonanResolvedModuleDescriptors(moduleDescriptors, forwardDeclarationsModule)
+    }
+
+    private fun createForwardDeclarationsModule(
+            builtIns: KotlinBuiltIns?,
```

Comment:
Would it be correct if `builtIns` was `null`? If no, then it is better to make its type non-nullable.

Reply:
I believe that if this function gets `builtIns = null` and then creates builtIns from forward declaration module, then the result will be incorrect, because it doesn't have stdlib in dependencies.

Reply:
Discussed outside. I will add more detailed comment explaining the danger of supplying `null` builtIns to this method.